### DISCUSSION TITLE
smartplaylists help

### DISCUSSION THREAD

**Discussion by mgonzalezm** [jcorporation/myMPD#1089] [Q&A] [answered]
Is it there a way to create a playlist that contains the list of songs that have never been played using the smartplaylists functionality ?

looking at the definition of the mostPlayed smartplaylist example, I tried to change minvalue:10 to maxvalue:0 , but that din't work.  :wink: :wink:

`myMPDsmart-mostPlayed: {"type": "sticker", "sticker": "playCount", "maxentries": 200, "minvalue": 10,"sort": ""}`

**Comment by jcorporation** [comment]
No, this is not implemented, but it could be easily implemented.

**Reply by mgonzalezm** [reply]
Thanks, that would be great

BTW, the above definition returns a <random?> selection of 200 songs or less from the universe of songs that have playCount >= 10, but when there is more tan 200 songs that meet that condition, it doesn't necessarily include the TOP played songs; those may be left out. 
Is that correct?

If that is the case, wouldn't be more proper to first sort the list, and then select from it. Something equivalent to this:

`select uri from sticker where type='song' and name='playCount' and value >= '10' order by value desc limit 200;`

**Reply by jcorporation** [reply, marked as answer]
myMPD includes the first 200 tracks that match the criteria. Sorting like you described is not possible through the mpd protocol.

myMPD v14 will focus on sticker handling enhancements. I add this to the milestone.

**Reply by mgonzalezm** [reply]
Great, I look forward to it

**Comment by jcorporation** [comment]
> Is it there a way to create a playlist that contains the list of songs that have never been played using the smartplaylists functionality ?

This is not possible. A song that have never been played has no sticker attached to it and the mpd protocol has no command to get songs without a specific sticker. We can only get songs that have a specific sticker attached to it.

**Reply by mgonzalezm** [reply]
right, that is too bad.

Thinking out loud here, what about if upon database update could the playCount be set to zero for the new songs.

Maybe using the trigger facility to run a script like this is enough. But i don't know the performance implications for those with huge libraries.

```
#!/bin/bash

shopt -s expand_aliases
alias mpc='mpc --host=/run/mpd.socket'

for f in $(mpc listall -f '%file%')
do
  mpc sticker "$f" get playCount &> /dev/null

  if [[ $? -eq 0 ]]
  then
    continue
  fi

  mpc sticker "$f" set playCount 0
done
```

**Reply by jcorporation** [reply]
> Thinking out loud here, what about if upon database update could the playCount be set to zero for the new songs.

This could be an option.

Your shell script can be optimized:

```
#!/bin/bash

shopt -s expand_aliases
alias mpc='mpc --host=/run/mpd.socket'

while read -r FILE
do
  if ! mpc sticker "$FILE" get playCount &> /dev/null
  then
    mpc sticker "$FILE" set playCount 0
 fi
done < <(mpc listall -f '%file%')
```

`for f in ...` reads tokens not lines and fails if there are blanks in the filename.

**Reply by mgonzalezm** [reply]
Yes that works for filenames with blanks, but with a big enough library and slow enough hardware, as it is in my case, it fails to process all of them.

It is so because the pipe operates in batches. It sends a batch of filenames and blocks while the loop is processing them, MPD doesn't like to wait for long (maybe because some output buffer fills up) and it drops the connection. The solution is to send the output of "mpc listall " to a file and read from there.

I will be running this shell triggered from the Update Database event, maybe it will be of some use once you drop v14 with new and exciting new stickers and smart-playlist functionality.

thank you

**Reply by jcorporation** [reply]
It could be the connection timeout. How big is your library?

**Reply by mgonzalezm** [reply]
about 25K.

The whole lot takes  2m 18s, when using the pipe it stop about half way.
